staging/src/k8s.io/apiserver/pkg/server/patch_genericapiserver.go

```diff
@@ -17,19 +17,97 @@ limitations under the License.
 package server
 
 import (
+	"fmt"
 	"net"
 	"net/http"
+	"os"
 	"strings"
 	"sync"
 	goatomic "sync/atomic"
+	"time"
 
 	"go.uber.org/atomic"
 
 	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/klog/v2"
 	netutils "k8s.io/utils/net"
 )
 
+// EventSink allows to create events.
+type EventSink interface {
+	Create(event *corev1.Event) (*corev1.Event, error)
+}
+
+type OpenShiftGenericAPIServerPatch struct {
+	// EventSink creates events.
+	eventSink EventSink
+	eventRef  *corev1.ObjectReference
+
+	// when we emit the lifecycle events, we store the event ID of the first
+	// shutdown event "ShutdownInitiated" emitted so we can correlate it to
+	// the other shutdown events for a particular apiserver restart.
+	// This provides a more deterministic way to determine the shutdown
+	// duration for an apiserver restart
+	eventLock                sync.Mutex
+	shutdownInitiatedEventID types.UID
+}
+
+// Eventf creates an event with the API server as source, either in default namespace against default namespace, or
+// if POD_NAME/NAMESPACE are set against that pod.
+func (s *GenericAPIServer) Eventf(eventType, reason, messageFmt string, args ...interface{}) {
+	t := metav1.NewMicroTime(time.Now())
+	host, _ := os.Hostname() // expicitly ignore error. Empty host is fine
+
+	ref := *s.eventRef
+	if len(ref.Namespace) == 0 {
+		ref.Namespace = "default" // TODO: event broadcaster sets event ns to default. We have to match. Odd.
+	}
+
+	e := &corev1.Event{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("%v.%x", ref.Name, t.UnixNano()),
+			Namespace: ref.Namespace,
+		},
+		EventTime:      t,
```

Comment:
Since we are creating corev1 Events, IMO we should mimic the behavior of the corev1 client implementation and set `FirstTimestamp` and `LastTimestamp` instead of `EventTime`: https://github.com/openshift/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/record/event.go#L482-L483

The events/v1 client implementation is actually the one setting EventTime: https://github.com/openshift/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/events/event_recorder.go#L102